File: dataStructure/hashTable/hashTable.js
const sha256 = require('crypto-js/sha256');

class Node {
  constructor(key, value) {
    this.key = key;
    this.value = value;
    this.prev = null;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
  }

  log() {
    let str = '';

    iterator = this.head;

    while (iterator instanceof Node) {
      str += `${iterator.key}: ${iterator.value} \n`;
      iterator = iterator.next;
    }

    return str;
  }

  findNodeWithKey(key) {
    let iterator = this.head;

    while (iterator instanceof Node) {
      if (iterator.key === key) {
        return iterator;
      }

      iterator = iterator.next;
    }

    return null;
  }

  push(key, value) {
    const newNode = new Node(key, value);

    if (this.head === null) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      newNode.prev = this.tail;
      this.tail.next = newNode;
      this.tail = newNode;
    }
  }

  delete(nodeToDelete) {
    if (nodeToDelete === this.head && nodeToDelete === this.tail) {
      this.head = null;
      this.tail = null;
      //
    } else if (nodeToDelete === this.head) {
      this.head = this.head.next;
      this.head.prev = null;
      //
    } else if (nodeToDelete === this.tail) {
      this.tail = this.tail.prev;
      this.tail.next = null;
      //
    } else {
      nodeToDelete.prev.next = nodeToDelete.next;
      nodeToDelete.next.prev = nodeToDelete.prev;
    }

    return nodeToDelete;
  }
}

class HashTable {
  constructor(capacity) {
    this._capacity = capacity;
    this._table = new Array(this._capacity).fill(new LinkedList());
  }

  log() {
    let str = '';

    for (let linkedList of this._table) {
      str += linkedList.log();
    }

    return str;
  }

  _hash(key) {
    return (
      sha256('영훈').words.reduce((pre, cur) => pre + cur) % this._capacity
    );
  }

  _getLinkedListForKey(key) {
    let hashIndex = this._hash(key);
    return this._table[hashIndex];
  }

  _lookUpNode(key) {
    let linkedList = this._getLinkedListForKey(key);
    return linkedList.findfindNodeWithKey(key);
  }

  lookUpValue(key) {
    return this._lookUpNode(key).value;
  }

  insert(key, value) {
    let node = this._lookUpNode(key);

    if (node) {
      node.value = value;
    } else {
      let linkedList = this._getLinkedListForKey(key);
      linkedList.push(key, value);
    }
  }

  deleteByKey(key) {
    let node = this._lookUpNode(key);
    let linkedList = this._getLinkedListForKey(key);

    if (node) {
      linkedList.delete(node);
      return node;
    }
  }
}

File: dataStructure/hashTable/hashTable.py
class Node:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def __str__(self):
        res_str = ""

        iterator = self.head

        while iterator is not None:
            res_str += "{}: {}\n".format(iterator.key, iterator.value)
            iterator = iterator.next

        return res_str

    def find_node_with_key(self, key):
        iterator = self.head
        
        while iterator is not None:
            if iterator.key == key:
                return iterator
            
            iterator = iterator.next
        
        return None

    def append(self, key, value):
        new_node = Node(key, value)

        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node

    def delete(self, node_to_delete):
        if node_to_delete is self.head and node_to_delete is self.tail:
            self.head = None
            self.tail = None
        
        elif node_to_delete is self.head:
            self.head = self.head.next
            self.head.prev = None

        elif node_to_delete is self.tail:
            self.tail = self.tail.prev
            self.tail.next = None
        
        else:
            node_to_delete.prev.next = node_to_delete.next
            node_to_delete.next.prev = node_to_delete.prev

        return node_to_delete

class HashTable:
    def __init__(self, capacity):
        self._capacity = capacity
        self._table = [LinkedList() for _ in range(self._capacity)]

    def __str__(self):
        res_str = ""

        for linked_list in self._table:
            res_str += str(linked_list)

        return res_str[:-1]

    def _hash_function(self, key):
        return hash(key) % self._capacity


    def _get_linked_list_for_key(self, key):
        hashed_index = self._hash_function(key)
        return self._table[hashed_index]


    def _look_up_node(self, key):
        linked_list = self._get_linked_list_for_key(key)
        return linked_list.find_node_with_key(key)

    def look_up_value(self, key):
        return self._look_up_node(key).value

            
    def insert(self, key, value):
        existing_node = self._look_up_node(key)

        if existing_node is not None:
            existing_node.value = value
        else:
            linked_list = self._get_linked_list_for_key(key)
            linked_list.append(key, value)

    def delete_by_key(self, key):
         existing_node = self._look_up_node(key)
         
         if existing_node is not None:
            linked_list = self._get_linked_list_for_key(key)
            linked_list.delete(existing_node)
            return existing_node

test_scores = HashTable(50)  # 시험 점수를 담을 해시 테이블 인스턴스 생성

# 여러 학생들 이름과 시험 점수 삽입
test_scores.insert("현승", 85)
test_scores.insert("영훈", 90)
test_scores.insert("동욱", 87)
test_scores.insert("지웅", 99)
test_scores.insert("신의", 88)
test_scores.insert("규식", 97)
test_scores.insert("태호", 90)

print(test_scores)

# key인 이름으로 특정 학생 시험 점수 검색
print(test_scores.look_up_value("현승"))
print(test_scores.look_up_value("태호"))
print(test_scores.look_up_value("영훈"))

# 학생들 시험 점수 수정
test_scores.insert("현승", 10)
test_scores.insert("태호", 20)
test_scores.insert("영훈", 30)

print(test_scores)
